### PR TITLE
[BE][Easy] export explicitly imported public submodules

### LINKED ISSUE #126401: TYP: declare torch.utils in __all__
### 🐛 Describe the bug

`torch.utils` is accessible but utils is not declared in `__all__` in `torch/.__init__.py` so type checkers do not know it exists.

```py
impor torch

torch.utils.data.Dataset  # pyright: "utils" is not a known attribute of module "torch" (reportAttributeAccessIssue)
```

### Versions

Collecting environment information...
PyTorch version: 2.3.0
Is debug build: False
CUDA used to build PyTorch: None
ROCM used to build PyTorch: N/A

OS: macOS 14.4.1 (arm64)
GCC version: Could not collect
Clang version: 15.0.0 (clang-1500.3.9.4)
CMake version: Could not collect
Libc version: N/A

Python version: 3.10.14 | packaged by conda-forge | (main, Mar 20 2024, 12:51:49) [Clang 16.0.6 ] (64-bit runtime)
Python platform: macOS-14.4.1-arm64-arm-64bit
Is CUDA available: False
CUDA runtime version: No CUDA
CUDA_MODULE_LOADING set to: N/A
GPU models and configuration: No CUDA
Nvidia driver version: No CUDA
cuDNN version: No CUDA
HIP runtime version: N/A
MIOpen runtime version: N/A
Is XNNPACK available: True

CPU:
Apple M2 Pro

Versions of relevant libraries:
[pip3] numpy==1.24.3
[pip3] onnx==1.16.0
[pip3] tf2onnx==1.16.1
[pip3] torch==2.3.0
[pip3] torchaudio==2.3.0
[conda] numpy                     1.24.3                   pypi_0    pypi
[conda] torch                     2.3.0                    pypi_0    pypi
[conda] torchaudio                2.3.0                    pypi_0    pypi

cc @ezyang @malfet @xuzhao9 @gramster @albanD

### PR DESCRIPTION
**Pull request by XuehaiPan**
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #127705
* #127704
* #128342
* #127707
* #127706
* #127710
* #127709
* #127708
* __->__ #127703

Add top-level submodules `torch.{storage,serialization,functional,amp,overrides,types}`

cc @albanD @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @chenyang78 @kadeng @chauhang

### CODE REVIEW

File: torch/utils/data/sampler.py [hunk 1]
```diff
@@ -212,7 +211,7 @@ class WeightedRandomSampler(Sampler[int]):
         [0, 1, 4, 3, 2]
     """
 
-    weights: Tensor
+    weights: 'torch.Tensor'
```
Comment (Skylion007) [inline torch/utils/data/sampler.py:214]: Why? Just making it harder for the type checker.

**Comment by Skylion007** [comment]
Some of these changes seem related to: https://github.com/pytorch/pytorch/pull/127688 and probably should wait on that.

File: torch/utils/data/sampler.py (hunk 1 above)
Comment (XuehaiPan) [reply torch/utils/data/sampler.py:214]: This is used to avoid cyclic imports. `torch.Tensor` is defined in `torch._tensor`.

```python
torch._tensor -> torch.utils.hooks -> torch.utils -> torch.utils.data
```

Comment (Skylion007) [reply torch/utils/data/sampler.py:214]: @XuehaiPan Just import in an `if TYPE_CHECKING` block then

File: torch/optim/__init__.py [hunk 2]
```diff
@@ -6,21 +6,21 @@
 future.
 """
 
-from . import lr_scheduler, swa_utils
```
Comment (Skylion007) [inline torch/optim/__init__.py:9]: Is ufmt forcing the removal of all these relative imports?

Comment (XuehaiPan) [reply torch/optim/__init__.py:9]: No. This is updated by hand.

> #### Absolute imports, or relative imports from siblings, are recommended by [PEP8](https://www.python.org/dev/peps/pep-0008/).

> Absolute imports are recommended, as they are usually more readable and tend to be better behaved (or at least give better error messages) if the import system is incorrectly configured (such as when a directory inside a package ends up on `sys.path`):
> 
> ```python
> import mypkg.sibling
> from mypkg import sibling
> from mypkg.sibling import example
> ```
> 
> However, explicit relative imports are an acceptable alternative to absolute imports, especially when dealing with complex package layouts where using absolute imports would be unnecessarily verbose:
> 
> ```python
> from . import sibling
> from .sibling import example
> Standard library code should avoid complex package layouts and always use absolute imports.
> ```

------

We can enforce this by adding the [ruff rule `TID` (`flake8-tidy-imports`)](https://docs.astral.sh/ruff/rules/#flake8-tidy-imports-tid) with:

```toml
[tool.ruff.lint.flake8-tidy-imports]
ban-relative-imports = "all"  # defaults to "parents"; see https://docs.astral.sh/ruff/settings/#lintflake8-tidy-imports
```

File: torch/utils/data/sampler.py (hunk 1 above)
Comment (XuehaiPan) [reply torch/utils/data/sampler.py]: It seems to be resolved by removing `torch.utils.hooks` in `torch._tensor`.

File: torch/optim/__init__.py (hunk 2 above)
Comment (Skylion007) [reply torch/optim/__init__.py:9]: Can't this affect import ordering which can be bad for our Python submodules in torch that do have side effects (or cause cyclic import errors where non existed before).

Comment (XuehaiPan) [reply torch/optim/__init__.py:9]: Changing relative import to identical absolute import does not change the side effects of the import statement.

File: torch/__init__.py [hunk 3]
```diff
@@ -565,20 +551,18 @@ def fn(a):
     sym_name = f"_sym_{name}"
     fn = _get_sym_math_fn(name)
     fn.__qualname__ = fn.__name__ = sym_name
-    setattr(current_module, sym_name, fn)
+    globals()[sym_name] = fn
```
Comment (XuehaiPan) [inline torch/__init__.py:554]: Use `globals()` here rather than accessing `sys,modules[__name__]`.

File: torch/optim/__init__.py (hunk 2 above)
Comment (albanD) [reply torch/optim/__init__.py:9]: Doesn't it change the import order though?
Where the relative import would actually trigger the import of the file while the absolute import would expect the partially initialized module to already contain the corresponding entry?

**Review by albanD — COMMENTED** [review]
This sounds good in principle but it shuffles a lot of APIs that, even though not public, might be used.

I am a bit concerned by this PR because:
- We change a lot of visible APIs in this one PR (even though they're technically not public API, we don't want to hard break anyone for no good reason), so there is a non-trivial risk that this will break a third party library and we will have to revert it.
- It is at the bottom of a pretty deep stack, meaning that reverting it is going to be quite disruptive.

We might want to split this into a PR that only change import to be explicit. And a different PR that cleans up other imports (like removing torch.utils.data.sampler.Tensor).
What do you think?

File: torch/optim/__init__.py (hunk 2 above)
Comment (XuehaiPan) [reply torch/optim/__init__.py:9]: This change is made based on:

1. The relative import will convert to absolute import in bytecode eval loop:

```console
$ echo 'from torch.optim import adam' | python3 -m dis -
  0           0 RESUME                   0

  1           2 LOAD_CONST               0 (0)
              4 LOAD_CONST               1 (('adam',))
              6 IMPORT_NAME              0 (torch.optim)
              8 IMPORT_FROM              1 (adam)
             10 STORE_NAME               1 (adam)
             12 POP_TOP
             14 RETURN_CONST             2 (None)
$ echo 'from . import adam' | python3 -m dis -
  0           0 RESUME                   0

  1           2 LOAD_CONST               0 (1)
              4 LOAD_CONST               1 (('adam',))
              6 IMPORT_NAME              0
              8 IMPORT_FROM              1 (adam)
             10 STORE_NAME               1 (adam)
             12 POP_TOP
             14 RETURN_CONST             2 (None)
```

https://github.com/python/cpython/blob/a9f2daf1ab182d95b44ee94dc9fb8faec60e34b1/Python/ceval.c#L2513-L2526

`from . import adam` is convert to `__import__(f'{__name__}.adam')` where `__name__ = 'torch.optim'`.
	
They are semantically identical.

2. See https://github.com/pytorch/pytorch/pull/127703#discussion_r1623617154: Absolute imports, or relative imports from siblings, are recommended by [PEP8](https://www.python.org/dev/peps/pep-0008/).

**Comment by XuehaiPan** [comment]
> We might want to split this into a PR that only change import to be explicit. And a different PR that cleans up other imports (like removing torch.utils.data.sampler.Tensor).

I moved the export for `torch.utils` to a separate PR:

- #128342

**Comment by XuehaiPan** [comment]
@Skylion007 could you please have a look at this? This PR is very easy to review after the split.

File: torch/__init__.py [hunk 4]
```diff
@@ -1666,7 +1654,9 @@ def _dtype(self):
 _tensor_classes: Set[Type] = set()
 
 # If you edit these imports, please update torch/__init__.py.in as well
+from torch import random as random
```
Comment (ezyang) [inline torch/__init__.py:1657]: is there a reason these have to live earlier here

File: torch/__init__.py [hunk 5]
```diff
@@ -1802,22 +1792,22 @@ def _assert(condition, message):
 from torch import nn as nn
 from torch.signal import windows as windows
 from torch import optim as optim
-import torch.optim._multi_tensor
```
Comment (ezyang) [inline torch/__init__.py:1805]: Hm?

Comment (ezyang) [reply torch/__init__.py:1805]: oh you restored it, okey

**Review by ezyang — APPROVED** [review]
I have audited the changes and it looks semantics preserving

**Comment by malfet** [comment]
@fbgheith reports that this broke some lazy import logic (still trying
 to figure out the details), so I'll be attempting to revert this and few others that depend on it to restore original behavior)

**Comment by XuehaiPan** [comment]
> @fbgheith reports that this broke some lazy import logic (still trying to figure out the details), so I'll be attempting to revert this and few others that depend on it to restore original behavior)

I will help with fixing that in PR #129095.

- #129095

File: torch/optim/__init__.py (hunk 2 above)
Comment (fbgheith) [reply torch/optim/__init__.py:9]: How about absolute paths when combined with circular dependencies? Example:

torch/optim/__init__.py import `torch.optim._multi_tensor` and
torch/optim/_multi_tenoser/__init__.py import `torch.optim`

This pattern, combined with absolute import paths is causing unpredictable and hard to debug failures.

**Comment by fbgheith** [comment]
This has been a big source of headache where depending on the exact ordering if imports in the pytorch client code one of three things would happen:

(1) great majority: things work fine
(2a) the "import torch.optim._multi_tensor" in torch/optim/__init__.py causes infinite recursion in a small set of uses cases
(2b) removing the import above solves (2a) but makes "import torch.optim._multi_tensor" fail for another small set
(2c) replacing the backwards absolute import with relative ones make both sets happy

This is just empirical data for this pair of dependencies.

**Comment by fbgheith** [comment]
Please keep in mind that #129095 is just a bandaid to remedy the particular scenario described above. We also need to come up with an explanation for why we ended up in this state and the proper guidelines for what can go in __init__.py files.

**Comment by XuehaiPan** [comment]
> We also need to come up with an explanation for why we ended up in this state and the proper guidelines for what can go in `__init__.py` files.

We will be able to do this after we enable `ufmt` globally and optionally enable `ruff`'s `TID` rules.
